FFmpeg video playback

feliwir opened the pull request:
Missing audio playback - this will probably only work with OpenAL. Adding Miles support would be more work

feliwir committed:
Add common class for audio & video decoding

feliwir committed:
Add FFmpeg video player & stream

feliwir committed:
Add a simple test that covers basic functionality

feliwir committed:
Add seeking support

Implement suggested changes

xezon commented on src/platform/audio/ffmpegaudiofilecache.cpp:
```diff
@@ -306,25 +151,23 @@ AudioDataHandle FFmpegAudioFileCache::Open_File(const Utf8String &filename)
     FFmpegOpenAudioFile open_audio;
```
perhaps call it `audio_file`. In several functions.

feliwir replied:
I was taking the naming from `MilesAudioFileCache`. Should i still change?

xezon replied:
Up to you. `open_audio` reads odd, because what is open about this audio? Would there be a `close_audio`? So I think `audio_file` would be a better name.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+        return false;
+    }
+
+    size_t avio_ctx_buffer_size = 0x10000;
```
constexpr

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    m_streams.resize(m_fmt_ctx->nb_streams);
+    for (unsigned int stream_idx = 0; stream_idx < m_fmt_ctx->nb_streams; stream_idx++) {
+        const AVCodec *input_codec = avcodec_find_decoder(m_fmt_ctx->streams[stream_idx]->codecpar->codec_id);
+        if (!input_codec) {
```
Style: `input_codec == nullptr`. Several times in functions.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    for (auto &stream : m_streams) {
+        if (stream.codec_ctx) {
```
Style: `!= nullptr`, several times in functions.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+            avcodec_free_context(&stream.codec_ctx);
+            av_frame_free(&stream.frame);
+        }
+    }
```
`m_streams` is not cleared.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+
+bool FFmpegFile::Decode_Packet()
+{
+    int result = av_read_frame(m_fmt_ctx, m_packet);
```
`Decode_Packet` is a public function. There should be tests or asserts on m_packet, m_streams & co.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+        captainslog_error("Failed 'avcodec_send_packet': %s", error_buffer);
+        return false;
+    }
+    av_packet_unref(m_packet);
```
What does this call pair to?

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+FFmpegFile::FFmpegStream *FFmpegFile::Find_Match(int type)
```
`AVMediaType type` ?

feliwir replied:
The public header only does forward declaration. `AVMediaType` is not available in the header

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    return stream->codec_ctx->pix_fmt;
+}
+
+} // namespace Thyme
```
eol

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+FFmpegFile::FFmpegStream *FFmpegFile::Find_Match(int type)
```
Looks as if it can return `const FFmpegFile::FFmpegStream`. And function can be const.

xezon commented on src/platform/common/ffmpegfile.h:
```diff
@@ -0,0 +1,80 @@
+    void Seek_Frame(int frame_idx);
+
+    // Audio specific
+    int Get_Size_For_Samples(int numSamples);
```
All these read-only functions should be `const`.

xezon commented on src/platform/common/ffmpegfile.h:
```diff
@@ -0,0 +1,80 @@
+        AVCodecContext *codec_ctx = nullptr;
+        const AVCodec *codec = nullptr;
+        int stream_idx = -1;
+        int stream_type = -1;
```
`AVMediaType stream_type = AVMEDIA_TYPE_UNKNOWN`

feliwir replied:
I can't do this, since the `AVMediaType` is not available inside the header

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+FFmpegFile::FFmpegStream *FFmpegFile::Find_Match(int type)
+{
+    for (auto &candidate : m_streams) {
+        if (candidate.stream_type == type)
```
Lookup is quite inefficient.

How about have a `std::array<FFmpegStream, 6>` and fill that? Then it can be indexed.

feliwir replied:
Usually an audiofile only has 1 stream and a bink video does have 2 streams. Also there can be multiple audio & video streams in theory (e.g. for different languages) - this function only finds the first match

xezon replied:
Ok not a big deal.

xezon commented on src/platform/video/ffmpegvideoplayer.cpp:
```diff
@@ -0,0 +1,119 @@
+    Utf8String path;
+
+    // Check for a language specific directoy Data/%language%/Movies.
+    path.Format("Data/%s/Movies/%s.%s", Get_Registry_Language().Str(), vid->file_name.Str(), "bik");
```
Why not look in mod directory first as it does in `BinkVideoPlayer::Open` ?

feliwir replied:
Will fix - must have missed it somehow

xezon commented on src/platform/video/ffmpegvideoplayer.cpp:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Uninitilise the class.
+ */
+void FFmpegVideoPlayer::Deinit() {}
```
Not call `VideoPlayer::Deinit()` here? BinkVideoPlayer::Deinit does it.

feliwir replied:
Will fix - must have missed it somehow

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    if (!stream)
+        return 0;
+
+    return stream->codec_ctx->channels;
```
`warning C4996: 'AVCodecContext::channels': was declared deprecated`

feliwir replied:
This is due to the fact that we "support" multiple ffmpeg versions. Some distros ship with `4.x`, others with `5.x` and others again with even `6.0`. In one place i even check for `LIBAVFORMAT_VERSION_MAJOR` to accomodate for that

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+namespace Thyme
+{
+
+SWVideoBuffer::SWVideoBuffer(Type type) : VideoBuffer(type), m_data(nullptr), m_locked(false) {}
```
`warning C5038: data member 'Thyme::SWVideoBuffer::m_data' will be initialized after data member 'Thyme::SWVideoBuffer::m_locked'`

xezon commented on src/platform/video/ffmpegvideostream.h:
```diff
@@ -0,0 +1,53 @@
+namespace Thyme
+{
+class FFmpegFile;
+class FFmpegVideoStream final : public VideoStream
```
```
29>ffmpegvideostream.obj : error LNK2019: unresolved external symbol _sws_freeContext referenced in function "public: virtual __thiscall Thyme::FFmpegVideoStream::~FFmpegVideoStream(void)" (??1FFmpegVideoStream@Thyme@@UAE@XZ)
29>ffmpegvideostream.obj : error LNK2019: unresolved external symbol _sws_scale referenced in function "public: virtual void __thiscall Thyme::FFmpegVideoStream::Render_Frame(class VideoBuffer *)" (?Render_Frame@FFmpegVideoStream@Thyme@@UAEXPAVVideoBuffer@@@Z)
29>ffmpegvideostream.obj : error LNK2019: unresolved external symbol _sws_getCachedContext referenced in function "public: virtual void __thiscall Thyme::FFmpegVideoStream::Render_Frame(class VideoBuffer *)" (?Render_Frame@FFmpegVideoStream@Thyme@@UAEXPAVVideoBuffer@@@Z)
```
On Windows.

feliwir replied:
I had that aswell IIRC. I think you need to reconfigure / delete the FFmpeg CMake variables when you're coming from the OpenAL branch. I added "SWSScale" as a new dependency, that somehow isn't picked up without doing what i described

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+
+FFmpegVideoStream::~FFmpegVideoStream()
+{
+    if (m_swsContext != nullptr) {
```
Obsolete test.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+        m_swsContext = nullptr;
+    }
+
+    if (m_ffmpegFile != nullptr) {
```
Obsolete test.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+    m_swsContext = sws_getCachedContext(m_swsContext,
+        Width(),
+        Height(),
+        (enum AVPixelFormat)m_frame->format,
```
static_cast

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    int dst_strides[] = { (int)buffer->Get_Pitch() };
+    uint8_t *dst_data[] = { buffer_data };
+    int result = sws_scale(m_swsContext, m_frame->data, m_frame->linesize, 0, Height(), dst_data, dst_strides);
```
How does this know size of arrays?

feliwir replied:
This depends on the pixelformat. E.g. YUV420P has multiple planes (Y and UV planes), while RGB32 everything is stored in a linear buffer. Since all pixelformats we convert to are linear, we're always have a single buffer. 
Formats are:
- AV_PIX_FMT_RGB24
- AV_PIX_FMT_BGR0
- AV_PIX_FMT_RGB565
- AV_PIX_FMT_RGB555

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+    return m_ffmpegFile->Get_Width();
+}
+
+} // namespace Thyme
```
eol

xezon commented on src/platform/video/ffmpegvideostream.h:
```diff
@@ -0,0 +1,53 @@
+    FFmpegFile *m_ffmpegFile = nullptr;
+    bool m_good = true;
+};
+} // namespace Thyme
```
eol

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+
+FFmpegVideoStream::FFmpegVideoStream(FFmpegFile *file) : m_ffmpegFile(file)
+{
+    file->Set_Frame_Callback(On_Frame);
```
Assert on `m_ffmpegFile != nullptr`, otherwise this class logic is malformed.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+{
+    if (m_swsContext != nullptr) {
+        sws_freeContext(m_swsContext);
+        m_swsContext = nullptr;
```
Obsolete assignment in destructor.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    if (m_ffmpegFile != nullptr) {
+        delete m_ffmpegFile;
+        m_ffmpegFile = nullptr;
```
Obsolete assignment in destructor.

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+    return m_data != nullptr;
+}
+
+} // namespace Thyme
```
eol

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+    int bytes_per_pixel = (m_format == VideoBuffer::TYPE_X8R8G8B8) ? 4 : 3;
+    m_pitch = bytes_per_pixel * m_width;
+    m_data = new uint8_t[m_pitch * m_height];
+    std::memset(m_data, 0, m_pitch * m_height);
```
You can save on this memset by writing:

`m_data = new uint8_t[m_pitch * m_height]();`

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void SWVideoBuffer::Free()
+{
+    if (m_data != nullptr) {
```
Obsolete test.

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+void *SWVideoBuffer::Lock()
+{
+    m_locked = true;
+    return m_data;
```
This function looks odd. It locks always regardless of prior locks.

feliwir replied:
Yes, it made no sense to me to have any "Locks" for the software video buffer. The locking concept makes sense for hardware buffers, but i didn't really know what to do for a memory buffer

xezon replied:
Perhaps remove `m_locked` if there is no locking?

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+    m_textureWidth = width;
+    m_textureHeight = height;
+
+    if (m_format != VideoBuffer::TYPE_X8R8G8B8 && m_format != VideoBuffer::TYPE_R8G8B8) {
```
This condition will always be true.

feliwir replied:
Why would it always be true? It's only true if it's not TYPE_X8R8G8B8 and not TYPE_R8G8B8

xezon replied:
Right. Misread condition.

xezon commented on src/platform/video/swvideobuffer.cpp:
```diff
@@ -0,0 +1,72 @@
+    m_textureWidth = width;
+    m_textureHeight = height;
+
+    if (m_format != VideoBuffer::TYPE_X8R8G8B8 && m_format != VideoBuffer::TYPE_R8G8B8) {
```
Perhaps move this check to the top of the function, so that nothing is changed if it will fail anyway.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,339 @@
+    if (result == AVERROR_EOF) // EOF
+        return false;
+
+    int stream_idx = m_packet->stream_index;
```
const

xezon commented on tests/test_videoplayer.cpp:
```diff
@@ -0,0 +1,55 @@
+    int count = stream->Frame_Count();
+    VideoBuffer *buffer = new Thyme::SWVideoBuffer(W3DVideoBuffer::TYPE_X8R8G8B8);
+    ASSERT_TRUE(buffer->Allocate(stream->Width(), stream->Height()));
+    while (idx < count) {
```
Can use `for (int idx = stream->Frame_Index(); idx < count; ++idx)` here to simplify it.

xezon commented on tests/test_videoplayer.cpp:
```diff
@@ -0,0 +1,55 @@
+    EXPECT_EQ(stream->Width(), 96);
+
+    int idx = stream->Frame_Index();
+    int count = stream->Frame_Count();
```
const

xezon commented on tests/test_videoplayer.cpp:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    delete buffer;
+    delete stream;
```
If I am not mistaken `delete stream` will cause dangling pointer on `m_firstStream` via `FFmpegVideoPlayer::Create_Stream`. In my opinion there should be a `FFmpegVideoPlayer::Destroy_Stream` function pair to properly clean up.

Also see `VideoPlayer::Remove_Stream`

feliwir replied:
I think destructor of `VideoStream` calls `m_player->Remove_Stream`, so this should be called

xezon replied:
Ok. Would it be possible to create a `FFmpegVideoPlayer::Destroy_Stream` function to pair `FFmpegVideoPlayer::Create_Stream`? It generally is bad practice to call `delete` on something that we did not call `new` on ourselves.

feliwir replied:
I was trying to stick with the `BinkVideoPlayer` / `VideoPlayer` interface. Inside "Stop_Movie" it calls `VideoStream::Close`, which just performs `delete this`. So ideally i'd replace this and use `->Close()`

xezon commented on src/platform/audio/ffmpegaudiofilecache.cpp:
```diff
@@ -55,194 +54,39 @@ FFmpegAudioFileCache::~FFmpegAudioFileCache()
 {
-    AVPacket *packet = av_packet_alloc();
-    AVFrame *frame = av_frame_alloc();
+    auto on_frame = [](AVFrame *frame, int stream_idx, int stream_type, void *user_data) {
```
This will be easier to debug if this is not a lambda function. Just make it a regular function in an anonymous namespace.

xezon commented on src/platform/audio/ffmpegaudiofilecache.cpp:
```diff
@@ -55,194 +54,39 @@ FFmpegAudioFileCache::~FFmpegAudioFileCache()
 
-    int result = 0;
-    int64_t total_samples = 0;
+        int frame_data_size = file->ffmpeg_file->Get_Size_For_Samples(frame->nb_samples);
```
const

feliwir committed:
Support multiple frames per packet

feliwir committed:
Start working on audio streans

Add FFmpeg video audio

Convert planar audio to interleaved

Support audio output in videoplayer test

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -159,42 +159,48 @@ int FFmpegFile::Read_Packet(void *opaque, uint8_t *buf, int buf_size)
+    m_streams.clear();
 
-    if (m_avio_ctx && m_avio_ctx->buffer) {
+    if (m_avio_ctx != nullptr && m_avio_ctx->buffer != nullptr) {
```
`m_avio_ctx->buffer != nullptr` is obsolete test.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -159,42 +159,48 @@ int FFmpegFile::Read_Packet(void *opaque, uint8_t *buf, int buf_size)
     }
 
-    if (m_avio_ctx) {
+    if (m_avio_ctx != nullptr) {
```
Is obsolete test.

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -159,42 +159,48 @@ int FFmpegFile::Read_Packet(void *opaque, uint8_t *buf, int buf_size)
     }
 
-    if (m_packet) {
+    if (m_packet != nullptr) {
```
Is obsolete test.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -18,12 +18,13 @@ FFmpegVideoStream::FFmpegVideoStream(FFmpegFile *file) : m_ffmpegFile(file)
     file->Set_User_Data(this);
     // Decode until we have our first video frame
-    while (m_good && m_frame == nullptr)
+    while (m_good && m_got_frame == false)
```
For false boolean test we typically use `!val`

2x in this file.

xezon commented on src/platform/audio/alaudiostream.cpp:
```diff
@@ -0,0 +1,37 @@
+    alBufferData(current_buffer, format, data, data_size, samplerate);
+    m_current_buffer_idx++;
+}
+} // namespace Thyme
```
eol

xezon commented on src/platform/audio/alaudiostream.h:
```diff
@@ -0,0 +1,38 @@
+    ALuint m_buffers[AL_STREAM_BUFFER_COUNT] = {};
+    unsigned int m_current_buffer_idx = 0;
+};
+} // namespace Thyme
```
eol

xezon commented on src/platform/audio/alaudiostream.h:
```diff
@@ -28,8 +28,15 @@ class ALAudioStream final
     ~ALAudioStream();
 
     bool BufferData(uint8_t *data, size_t data_size, ALenum format, int samplerate);
+    bool IsPlaying();
```
Can be const?

xezon commented on src/platform/video/ffmpegvideoplayer.cpp:
```diff
@@ -34,6 +38,7 @@ FFmpegVideoPlayer::~FFmpegVideoPlayer()
 void FFmpegVideoPlayer::Init()
 {
     VideoPlayer::Init();
+    Initialise_FFmpeg_With_OpenAL();
```
Engine is create by Americans. Americans use "Initialize". Several times in this code.

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -35,24 +43,46 @@ void FFmpegVideoStream::On_Frame(AVFrame *frame, int stream_idx, int stream_type
+    }
+#ifdef BUILD_WITH_OPENAL
+    else if (stream_type == AVMEDIA_TYPE_AUDIO) {
+        FFmpegVideoPlayer *player = static_cast<FFmpegVideoPlayer *>(stream->m_player);
```
video_player

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,249 @@
+
+void FFmpegVideoStream::On_Frame(AVFrame *frame, int stream_idx, int stream_type, void *user_data)
+{
+    FFmpegVideoStream *stream = static_cast<FFmpegVideoStream *>(user_data);
```
video_stream

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -35,24 +43,46 @@ void FFmpegVideoStream::On_Frame(AVFrame *frame, int stream_idx, int stream_type
 {
-    // BinkWait(m_binkHandle);
+#ifdef BUILD_WITH_OPENAL
+    FFmpegVideoPlayer *player = static_cast<FFmpegVideoPlayer *>(m_player);
```
video_player

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -35,24 +43,46 @@ void FFmpegVideoStream::On_Frame(AVFrame *frame, int stream_idx, int stream_type
-    // BinkWait(m_binkHandle);
+#ifdef BUILD_WITH_OPENAL
+    FFmpegVideoPlayer *player = static_cast<FFmpegVideoPlayer *>(m_player);
+    ALAudioStream *stream = player->Get_Audio();
```
audio_stream

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -131,9 +161,18 @@ void FFmpegVideoStream::Render_Frame(VideoBuffer *buffer)
 {
-    m_got_frame = false;
+#ifdef BUILD_WITH_OPENAL
+    FFmpegVideoPlayer *player = static_cast<FFmpegVideoPlayer *>(m_player);
```
video_player

xezon commented:
bik video and audio works.

The EA_LOGO.BIK playback is stuttering considerably.

The sizzle_review.bik playback is stuttering a bit and the audio is not in sync with the video.

feliwir committed:
Flush the audiostream when a new video starts

xezon commented:
EA_LOGO.BIK playback looks and sounds ok now.

But sizzle_review.bik and campaign videos no longer play audio at all.

feliwir committed:
Use an audiostream per videostream

xezon commented on src/platform/audio/alaudiostream.cpp:
```diff
@@ -0,0 +1,97 @@
+    alGenSources(1, &m_source);
+    alGenBuffers(AL_STREAM_BUFFER_COUNT, m_buffers);
+
+    // Make strream ignore positioning
```
stream

xezon commented on src/platform/audio/alaudiostream.h:
```diff
@@ -0,0 +1,46 @@
+
+protected:
+    ALuint m_source = 0;
+    ALuint m_buffers[AL_STREAM_BUFFER_COUNT] = {};
```
Perhaps use std::array and remove `AL_STREAM_BUFFER_COUNT`.

feliwir replied:
IIRC we couldn't use `std:array`. However we have a convenience `Array` class in `array.h`

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,363 @@
+            return false;
+        }
+
+        m_streams[stream_idx].codec_ctx = codec_ctx;
```
Can also introduce reference here:

```c++
FFmpegStream &output_stream = m_streams[stream_idx];
```

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,363 @@
+
+    m_streams.resize(m_fmt_ctx->nb_streams);
+    for (unsigned int stream_idx = 0; stream_idx < m_fmt_ctx->nb_streams; stream_idx++) {
+        const AVCodec *input_codec = avcodec_find_decoder(m_fmt_ctx->streams[stream_idx]->codecpar->codec_id);
```
Can also introduce reference here:

```c++
AVStream *av_stream = m_fmt_ctx->streams[stream_idx];
```

xezon commented on src/platform/common/ffmpegfile.cpp:
```diff
@@ -0,0 +1,363 @@
+    const FFmpegStream *stream = Find_Match(AVMEDIA_TYPE_VIDEO);
+    if (stream == nullptr)
+        return 0;
+    return 1000 / av_q2d(m_fmt_ctx->streams[stream->stream_idx]->avg_frame_rate);
```
1000u

xezon commented on src/platform/video/ffmpegvideoplayer.cpp:
```diff
@@ -0,0 +1,143 @@
+ */
+VideoStream *FFmpegVideoPlayer::Open(Utf8String title)
+{
+    Video *vid = Get_Video(title);
```
video

xezon commented on src/platform/video/ffmpegvideostream.cpp:
```diff
@@ -0,0 +1,247 @@
+            break;
+        case VideoBuffer::TYPE_X1R5G5B5:
+            dst_pix_fmt = AV_PIX_FMT_RGB555;
+            return;
```
break?

xezon reviewed (commented):
Bink video and its audio playback works.

xezon reviewed (approved):
Use squash and Merge.
